config/v1/0000_10_config-operator_01_authentication.crd.yaml

```diff
@@ -64,7 +64,7 @@ spec:
                   type: string
             serviceAccountIssuer:
               description: serviceAccountIssuer is the identifier of the bound service
-                account token issuer. The default is auth.openshift.io.
+                account token issuer. The default is https://apiserver.openshift-kube-apiserver.svc
```

Comment:
what does this mean for old existing service accounts?

how is this name used by kube?

Reply:
Kube defaults to using the dns for its apiserver - `kubernetes.default.svc`

This field was introduced in 4.4, which is not GA yet. The goal is updating the default and backporting to 4.4 before any customer has production clusters deployed with the old default (as per the targeted bug).

Note that the change here is just a comment. I'm in the process of validating the functional change to kao and will post shortly.

Reply:
In terms of how it's used by kube, it is only included in bound tokens as the issuer in 1.17/4.4. In 1.18, the ServiceAccountIssuerDiscovery feature is alpha and if enabled the apiserver will serve an oidc document and the keys needed to verify service account tokens as per [1]. By setting the default to the API server in openshift on introduction of bound token support in 4.4, this discovery feature can be enabled in the future for internal use without requiring a change to issuer (and the reissue of all bound tokens that implies).

1: https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-issuer-discovery

Reply:
could we default to kubernetes.default.svc as upstream?

Reply:
Upstream points to `https://kubernetes.default.svc` because that's the url for the api server in a kube cluster. When the ServiceAccountIssuerDiscovery feature is enabled, a recipient of a bound token can connect to apiserver identified in the embedded issuer url to retrieve an oidc and jwks (all the service account keys) docs.  

In openshift, the equivalent url is `https://apiserver.openshift-kube-apiserver.svc` since that's the service under which the apiserver is deployed. Specifying the same url as upstream would mean the recipient of a bound token wouldn't be able to use the issuer url to retreive the oidc document and jwks doc. Mind you, it may be that this default is really only good for testing since internal clients could as easily verify tokens with the TokenReview api.

@deads2k I remember you had concerns about setting issuer to a real url in the enhancement, what are your thoughts in light of the addition of the alpha ServiceAccountIssuerDiscovery feature in 1.18?

Reply:
I don't understand. We also have `https://kubernetes.default.svc`.

Reply:
I did not find a specific mention that this should point to `https://kubernetes.default.svc`, the only mention is that API servers are not usually publicly resolvable.

If we set it the this value to `https://kubernetes.default.svc`, we would by default be able to provide the OIDC discovery for in-cluster components. Per the motivation at https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/20190730-oidc-discovery.md#motivation, the intention rather is to allow SA authentication to out-of-cluster services without having to do TokenReviews.

Now we could say that `https://kubernetes.default.svc` is good enough as it at least won't break the SA Issuer discovery once turned on, and if a user wants to use that feature, they need to configure the field to point to a URL which resolves to KAS for the out-of-cluster service they mean to use it at. And this might be good enough.

Or, we could right away set it to the URL that we expose the KAS at, which would mean that it could be usable for external authentication right away (I think authenticating workloads to the underlying cloud provider would need that, too, no?).

I am not sure what would be needed to implement the latter solution, whether we have enough information in KAS-o when starting the static pods for the first time and whether something would have to be done on the bootstrap plane, or if that's even possible?

Reply:
> @deads2k I remember you had concerns about setting issuer to a real url in the enhancement, what are your thoughts in light of the addition of the alpha ServiceAccountIssuerDiscovery feature in 1.18?

I remember it being the case that we didn't know the actual resolvable name and that if we set it something that can resolve externally we would end up potentially mis-directing clients in a non-obvious way.  

We do want to be compatible with the ServiceAccountIssuerDiscovery when it comes.  Setting to an internal name as you've done seems reasonable.  In this case though, the kube-apiserver actually manually creates a kubernetes.default.svc service, so we can actually use the upstream as sttts noted.